packages/tuta-wasm-loader/lib/index.ts

```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+function esbuildWasmLoader(options: LoadOptions) {
+	return {
+		name: "wasm",
+		setup(build: PluginBuild) {
+			createOutputFolderStructure(options.output)
+
+			build.onResolve({ filter: /\.(?:wasm)$/ }, async (args): Promise<OnResolveResult | undefined> => {
+				if (args.resolveDir === "" && !args.path.includes("wasm-fallback")) return
+
+				if (args.path.includes("wasm-fallback")) {
+					return {
+						path: args.path.replaceAll("wasm-fallback:", ""),
+						namespace: "wasm-fallback",
+					}
+				}
+
+				const lib = findLib(options.webassemblyLibraries, args.path)
+
+				await generateWasm(lib.command, lib.options)
+
+				return {
+					path: path.join("wasm", args.path),
+					namespace: "wasm-loader",
+				}
+			})
+
+			build.onResolve({ filter: /\.(?:wasm-fallback)$/ }, async (args): Promise<OnResolveResult | undefined> => {
+				if (args.resolveDir === "") return
+				return {
+					path: args.path,
+					namespace: "wasm-fallback",
+				}
+			})
+
+			build.onLoad({ filter: /.*/, namespace: "wasm-loader" }, async (args): Promise<OnLoadResult> => {
+				return {
+					contents: await generateImportCode(args.path),
+					loader: "js",
+				}
+			})
+
+			build.onLoad({ filter: /.*/, namespace: "wasm-fallback" }, async (args): Promise<OnLoadResult> => {
+				const buildPath = options.output
+				const lib = findLib(options.webassemblyLibraries, args.path)
+				return {
+					contents: await generateWasmFallback(path.join(buildPath, args.path), {
+						optimizationLevel: lib.options.optimizationLevel,
+						env: lib.options.env,
+						wasm2jsPath: lib.options.wasm2jsPath,
+					}),
+					loader: "js",
+				}
+			})
+		},
+	}
+}
+
+function rollupWasmLoader(options: LoadOptions & { output: string }) {
+	createOutputFolderStructure(options.output)
+
+	return {
+		name: "wasm",
+		async resolveId(source: string) {
+			if (source.endsWith(".wasm")) {
+				return { id: `\0wasm-loader:${source}`, external: false }
+			}
+		},
+		async resolveDynamicImport(specifier: string, importer: string) {
+			if (importer.includes("wasm-loader")) {
+				return {
+					id: `\0${specifier}`,
+					external: false,
+				}
+			}
+		},
+		async load(id: string) {
+			if (id.startsWith("\0wasm-loader")) {
+				const wasmLib = id.replaceAll("\0wasm-loader:", "")
+
+				const lib = findLib(options.webassemblyLibraries, wasmLib)
+				await generateWasm(lib.command, lib.options)
+
+				return await generateImportCode(path.join("wasm", wasmLib))
+			} else if (id.startsWith("\0wasm-fallback")) {
+				const wasmLib = id.replaceAll("\0wasm-fallback:", "")
+				const lib = findLib(options.webassemblyLibraries, wasmLib)
+				const wasmPath = path.join(options.output, wasmLib)
+
+				return await generateWasmFallback(wasmPath, {
+					optimizationLevel: lib.options.optimizationLevel,
+					env: lib.options.env,
+					wasm2jsPath: lib.options.wasm2jsPath,
+				})
+			}
+		},
+		async transform(source: string, id: string) {
+			if (id.includes(".wasm")) {
```

Comment:
why is this part needed? sorry, I probably asked already before but better to add a comment

Reply:
Returning the generated loader or fallback. It's called after load and should return the JS code if the Id is already generated/imported or call **load** again to handle imports

https://rollupjs.org/plugin-development/#build-hooks

Reply:
Please add this comment to the code

Reply:
Done

Reply:
Okay damn I didn't read it properly, I didn't want to know what `transform()` does, I can look it up, I wanted to know why we need that condition for wasm files. Is it so that rollup doesn't try to transform them?

Reply:
> Okay damn I didn't read it properly, I didn't want to know what `transform()` does, I can look it up, I wanted to know why we need that condition for wasm files. Is it so that rollup doesn't try to transform them?

I check if it includes wasm in the name to prevent our plugin from processing non-wasm files. All project files goes through the plugin lifecycle (resolveId, load and transform) and I just ignore the ones that isn't related to WebAssembly stuff.

Reply:
but you always return `null` so you never process anything, why do we need to add this callback at all? You can just remove `transform` if i get it right? https://rollupjs.org/plugin-development/#a-simple-example

Reply:
So, during the rollup I replace the wasm call by **wasm-loader:\<lib\>.wasm** and I want to return the source code for the loader inside the transform, that's the reason for doing this if check, returning the source code **ONLY** for loader (**wasm-loader:\<lib\>.wasm**) and fallback (**wasm-fallback:\<lib\>.wasm**) and not for all the other files. Sorry for the confusion